Update hypershift docs + add tshooting (#2092)
timescale/docs: _troubleshooting/cloud/hypershift-bad-perf.md

---

> title: Hypershift migration runs slowly

Reviewer: Unfortunately there are multiple issues that manifest from this failure mode, not just "slowness".

What we want hypershift to do is to copy multiple parts of the source table in parallel (essentially chunks, but plain postgres doesn't have chunks). When compression is enabled on the destination table, we can compress each chunk immediately after copying, this is how we avoid peak storage usage.

Without an index, this parallel copy is not performed. This means that not only is the copy slower (because we couldn't copy in parallel), but the target DB instance will require the same amount of storage for the table as the source database did, at least until compression is applied.

I know this is a lot, and goes kind of deep into how Timescale and Hypershift works, but IMHO it's important that we explain why these effects happen.

Author: OK, I'm happy to include more detail in the entry, but we can probably leave the title as is?

Reviewer: Ya, sure, that works for me!

---
title: Hypershift migration runs slowly
section: troubleshooting
products: [cloud]
keywords: [migration, hypershift]
tags: [hypershift, migration, ingest, postgresql]
---

<!---
* Use this format for writing troubleshooting sections:
 - Cause: What causes the problem?
 - Consequence: What does the user see when they hit this problem?
 - Fix/Workaround: What can the user do to fix or work around the problem? Provide a "Resolving" Procedure if required.
 - Result: When the user applies the fix, what is the result when the same action is applied?
* Copy this comment at the top of every troubleshooting page
-->

When you migrate a PostgreSQL database to Timescale using the Hypershift tool,
the migration can run very slowly if you do not have an index on the `time`
column. Hypershift uses the index to determine how to perform the migration. It
uses the index to break the database into parts that is can copy in parallel.
Additionally, if you have compression enabled on the destination table,
Hypershift can compress each part immediately after it is copied, which uses

> less storage dueing the migration. Without an index, Hypershift attempts to copy

Reviewer: ```suggestion
less storage during the migration. Without an index, Hypershift attempts to copy
```

the entire table at once instead. To resolve this problem, create a btree index
on the `time` column. However, creating an index can take some time, as the
entire table needs to be read from disk. You can create the btree index with
this command:

```sql
CREATE INDEX ON "<TABLE_NAME>" USING btree (time);
```

<highlight type="important">

> Hypershift does not support composite indexes. Ensure your source database has
> a plain index before you run the Hypershift migration.

Reviewer: I'm a little concerned with the phrasing "does not support", it gives the impression that it might not be possible to use Hypershift at all when there is a composite index anywhere.

```suggestion
Hypershift is not able to efficiently copy and compress data when the only
index is a composite index where `time` is not the first indexed column. If you
already have such a composite index, ensure that your source database has a
plain index before you run the Hypershift migration.
```

---
title: Hypershift migration runs slowly
section: troubleshooting
products: [cloud]
keywords: [migration, hypershift]
tags: [hypershift, migration, ingest, postgresql]
---

<!---
* Use this format for writing troubleshooting sections:
 - Cause: What causes the problem?
 - Consequence: What does the user see when they hit this problem?
 - Fix/Workaround: What can the user do to fix or work around the problem? Provide a "Resolving" Procedure if required.
 - Result: When the user applies the fix, what is the result when the same action is applied?
* Copy this comment at the top of every troubleshooting page
-->

When you migrate a PostgreSQL database to Timescale using the Hypershift tool,
the migration can run very slowly if you do not have an index on the `time`
column. Hypershift uses the index to determine how to perform the migration. It

> uses the index to break the database into parts that is can copy in parallel.

Reviewer 2: ```suggestion
uses the index to break the database into parts that it can copy in parallel.
```

Additionally, if you have compression enabled on the destination table,
Hypershift can compress each part immediately after it is copied, which uses
less storage during the migration. Without an index, Hypershift attempts to copy

> the entire table at once instead. To resolve this problem, create a btree index

Reviewer 2: ```suggestion
the entire table at once instead. To resolve this problem, create a `btree` index
```

on the `time` column. However, creating an index can take some time, as the

> entire table needs to be read from disk. You can create the btree index with

Reviewer 2: ```suggestion
entire table needs to be read from disk. You can create the `btree` index with
```

this command:

```sql
CREATE INDEX ON "<TABLE_NAME>" USING btree (time);
```

> <highlight type="important">

Reviewer 2: ```suggestion
<Highlight type="important">
```

Author: One day I will start remembering this 🤦‍♀️

Hypershift is not able to efficiently copy and compress data when the only
index is a composite index where `time` is not the first indexed column. If you
already have such a composite index, ensure that your source database has a
plain index before you run the Hypershift migration.

> </highlight>

Reviewer 2: ```suggestion
</Highlight>
```
